### PR TITLE
Registration library

### PR DESCRIPTION
**Pull request by shuzhang1989**
it's used by rocksdb host to register to controller cluster manager

### CODE REVIEW

File: common/controller_proxy.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+// Proxy functions to talk to controller for cluster management functions.
+//
+// @author shu (shu@pinterest.com)
+//
+
+#pragma once
+
+#include <string>
+
+#include <gflags/gflags.h>
```
Comment (newpoo) [inline common/controller_proxy.h]: move it to .cpp file?

File: common/controller_proxy.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace common {
+
+std::string construct_controller_curl_cmd(
```
Comment (newpoo) [inline common/controller_proxy.h]: do we expect client use this function?
If not, let's move it to .cpp, and in test file, we can add a declaration there.

File: common/controller_proxy.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        const std::string& controller_http_curl,
+        const std::string& cluster_namespace,
+        const std::string& cluster_name,
+        const std::string& ip_string,
+        const int port, const std::string& az_string);
+
+// We need to specify port here because port flag is defined in realpin and
+// rocksplicator, so we avoid using gflag directly.
+bool registerHostToController(const int port);
```
Comment (newpoo) [inline common/controller_proxy.h]: use uint16_t?

File: common/controller_proxy.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+namespace common {
+
+std::string construct_controller_curl_cmd(
+        const std::string& controller_http_curl,
+        const std::string& cluster_namespace,
+        const std::string& cluster_name,
+        const std::string& ip_string,
+        const int port, const std::string& az_string);
+
+// We need to specify port here because port flag is defined in realpin and
```
Comment (newpoo) [inline common/controller_proxy.h]: add a few words describing what does this function do?

File: common/controller_proxy.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+
+//
+// @author shu (shu@pinterest.com)
+//
+
+#include "common/availability_zone.h"
```
Comment (newpoo) [inline common/controller_proxy.cpp]: sort the headers by following our conventions?
1. controller_proxy.h
2. c headers
3. c++ headers
4. other headers

File: common/controller_proxy.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    LOG(ERROR) << "Failed to popen() with errno: " << errno;
+    return false;
+  }
+
+  SCOPE_EXIT {
+    pclose(f);
+  };
+
+  char buf[128];
+  fread(buf, 1, sizeof(buf), f);
```
Comment (newpoo) [inline common/controller_proxy.cpp]: what do we expect the controller return?

File: common/controller_proxy.cpp [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        const std::string& cluster_name,
+        const std::string& ip_string, const int port,
+        const std::string& az_string) {
+  std::string mutate_ip_string = ip_string;
+  std::replace(mutate_ip_string.begin(), mutate_ip_string.end(), '.', '-');
+  std::string full_host_string = folly::stringPrintf(
+          "%s-%s-%s", mutate_ip_string.c_str(), std::to_string(port).c_str(),
+          az_string.c_str());
+
+  std::string full_curl_cmd =folly::stringPrintf(
```
Comment (newpoo) [inline common/controller_proxy.cpp]: one space after =

File: controller/tools/mysql/create_tables.sql [hunk 8]
```diff
@@ -14,8 +14,6 @@ CREATE TABLE IF NOT EXISTS tag (
   locks TINYINT UNSIGNED NOT NULL,
   created_at DATETIME NOT NULL,
   owner VARCHAR(256),
-  live_hosts MEDIUMTEXT,
```
Comment (newpoo) [inline controller/tools/mysql/create_tables.sql]: are these intended?

File: common/controller_proxy.h (hunk 1 above)
Comment (shuzhang1989) [reply common/controller_proxy.h]: sounds good

File: common/controller_proxy.h (hunk 2 above)
Comment (shuzhang1989) [reply common/controller_proxy.h]: sure, sounds good.

File: common/controller_proxy.h (hunk 4 above)
Comment (shuzhang1989) [reply common/controller_proxy.h]: sure

File: common/controller_proxy.h (hunk 3 above)
Comment (shuzhang1989) [reply common/controller_proxy.h]: port in realpin / rocksplicator are both decalred as glfag_int32, should we do the implicit conversion here?

File: common/controller_proxy.cpp (hunk 5 above)
Comment (shuzhang1989) [reply common/controller_proxy.cpp]: sure

File: common/controller_proxy.cpp (hunk 7 above)
Comment (shuzhang1989) [reply common/controller_proxy.cpp]: sure

File: common/controller_proxy.cpp (hunk 6 above)
Comment (shuzhang1989) [reply common/controller_proxy.cpp]: it should always return true

File: common/controller_proxy.h (hunk 3 above)
Comment (newpoo) [reply common/controller_proxy.h]: I slightly prefer using uint16_t here.

File: common/controller_proxy.cpp [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+#include "common/availability_zone.h"
+#include "common/network_util.h"
+
+#include <folly/ScopeGuard.h>
+#include <folly/String.h>
+#include <gflags/gflags.h>
+#include <glog/logging.h>
+
+#include <cstdio>
+#include <algorithm>
```
Comment (ywang282) [inline common/controller_proxy.cpp]: swap the order of <cstdio> and  <algorithm>?

File: common/controller_proxy.h [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+// Every Rocksplicator process will call this function before serve(),
+// this is to contact controller-http service for the existence of the host
+// in the cluster. Controller cluster management module will be able to
+// use the reported information to do cluster management tasks for example
+// replacing bad hosts.
+//
+// We need to specify port here because port flag is defined in realpin and
+// rocksplicator, so we avoid using gflag directly.
+bool registerHostToController(const uint16_t port);
```
Comment (angxu) [inline common/controller_proxy.h]: nit: capital "r" for registerHostToController?

File: common/controller_proxy.cpp (hunk 9 above)
Comment (shuzhang1989) [reply common/controller_proxy.cpp]: sure

File: common/controller_proxy.cpp (hunk 7 above)
Comment (newpoo) [reply common/controller_proxy.cpp]: add one space pls.

File: common/controller_proxy.h (hunk 10 above)
Comment (shuzhang1989) [reply common/controller_proxy.h]: ok

File: common/controller_proxy.h [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+//
+// Proxy functions to talk to controller for cluster management functions.
+//
+// @author shu (shu@pinterest.com)
+//
+
+#pragma once
+
+#include <string>
```
Comment (newpoo) [inline common/controller_proxy.h]: remove it?

**Pull request merged**